lib/1.4/dml-builtins.dml

```diff
@@ -578,6 +594,191 @@ template device {
     method destroy() default { }
 }
 
+/**
+## Device templates
+
+As the top-level scope defines the device object, device templates are
+templates intended to be instantiated on the top-level. Such templates are
+typically written in order to describe logic which affects the entire device
+model as a whole.
+
+The only built-in device template (save for the `device` template which defines
+the object type itself) is the `_thread_aware` template; however, the DML
+standard library (`utility.dml`) also defines the device templates [`poreset`,
+`hreset`, and `sreset`](utility.html#templates-for-reset), which declare device
+support for different kinds of resets.
+
+The `_thread_aware` template is offered as a means of writing *thread-aware* DML
+device models, in the sense defined by the *Simics API Reference Manual* (in
+section 2.7). This can be necessary in a multicore-acceleration setting, if the
+models would otherwise present a bottleneck in the simulation. The
+`_thread_aware` template **must be** and **should only be** used for this
+purpose.
+
+Instantiating the `_thread_aware` template on top-level declares the device
+model as being thread-aware, which carries a large number of ramifications and
+responsibilities.
+
+The consequences of instantiating the `_thread_aware` template are as follows:
+* The [`_thread_aware` parameter](#device-objects) of the device will resolve
+  to constant `true` rather than `false`. This can be leveraged by model code:
+  `dev._thread_aware` is suitable for use together with `#if` to either
+  conditionally generate additional code needed to support thread-aware device
+  models, or to conditionally raise compile-time errors via the [`error`
+  statement/declaration](language.html#error-statements) if some piece of code
+  does not support thread-aware device models.
+* An implementation of the `concurrency_mode` interface is provided for the
+  device object, together with an implementation of the `concurrency_group`
+  interface to explicitly group all ports, banks, and subdevices together
```

Comment:
Why not? I very much doubt that'll be particularly expensive. Especially since the `concurrency_group` interface should only be invoked once, after-configuration-but-before-sim-start, as the MCA mode is set-up, right?

Reply:
Double-checked. `VT_update_thread_domain_assignments()` is what invokes `concurrency_group` and forms thread domains. It's called as part of `add_execute_objects_to_cell`, which is documented as:
```
/* add clocks to the cell; this occurs either at finalize or when
   a clock has been moved to a new cell */
```
After that `concurrency_group` is not invoked. So performance is not a concern.

Reply:
good